core/sql/optimizer/BindRelExpr.cpp

```diff
@@ -10605,43 +10605,62 @@ NABoolean Insert::isUpsertThatNeedsTransformation(NABoolean isAlignedRowFormat,
-      (getTableDesc()->hasSecondaryIndexes() ))
-        {
-          toMerge = FALSE;
-          return TRUE;
-        }
-  else if (isUpsert() &&   (NOT getIsTrafLoadPrep()) && 
-           (NOT (getTableDesc()->isIdentityColumnGeneratedAlways() && getTableDesc()->hasIdentityColumnInClusteringKey())) && 
-           (NOT (getTableDesc()->getClusteringIndex()->getNAFileSet()->hasSyskey())) &&         
-           // CQD is set to MERGE  
-           ((CmpCommon::getDefault(TRAF_UPSERT_MODE) == DF_MERGE) &&
-            // omitted current default columns with non-aligned row format tables
-            // or omitted default columns with aligned row format tables 
-            (((NOT isAlignedRowFormat) && omittedCurrentDefaultClassCols) ||
-            (isAlignedRowFormat && omittedDefaultCols))) ||
-           // CQD is set to Optimal, for non-aligned row format with omitted 
-           // current columns, it is converted into merge though it is not
-           // optimal for performance - This is done to ensure that when the 
-           // CQD is set to optimal, non-aligned format would behave like 
-           // merge when any column is  omitted 
-           ((CmpCommon::getDefault(TRAF_UPSERT_MODE) == DF_OPTIMAL) &&
-            ((NOT isAlignedRowFormat) && omittedCurrentDefaultClassCols)))
+            getTableDesc()->hasIdentityColumnInClusteringKey()))  && 
+      (NOT(getTableDesc()->getClusteringIndex()->getNAFileSet()->hasSyskey())))
+    {
+      mustTryTransform = TRUE;
+    }
+
+  // Transform upsert to merge in case of special modes and
+  // omitted default columns
+  // Case 1 :  CQD is set to MERGE, omitted current(timestamp) default 
+  //           columns with  non-aligned row format table or omitted 
+  //           default columns with aligned row format tables 
+
+  // Case 2 :  CQD is set to Optimal, for non-aligned row format with omitted 
+  //           current(timestamp) columns, it is converted into merge 
+  //           though it is not optimal for performance. This is done to ensure
+  //           that when the CQD is set to optimal, non-aligned format would 
+  //           behave like merge when any column is  omitted 
+  if (isUpsert()  &&   
+      mustTryTransform &&          
+      ((CmpCommon::getDefault(TRAF_UPSERT_MODE) == DF_MERGE) &&     
+       (((NOT isAlignedRowFormat) && omittedCurrentDefaultClassCols) ||
+        (isAlignedRowFormat && omittedDefaultCols)))
+      ||
+      ((CmpCommon::getDefault(TRAF_UPSERT_MODE) == DF_OPTIMAL) &&
+       ((NOT isAlignedRowFormat) && omittedCurrentDefaultClassCols)))
     {
       toMerge = TRUE;
       return TRUE;
     }
-  toMerge = FALSE;
+
+  // Transform upsert to efficient tree if none of the above conditions 
+  // are true and the table has secondary indexes 
+  if (isUpsert() &&  
+      mustTryTransform &&
+      (getTableDesc()->hasSecondaryIndexes()))
+    {
+      toMerge = FALSE;
+      return TRUE;
+    }
+  
```

Comment:
@sureshsubbiah @zellerh  Please take a look at this file's rework and see if it looks better. Also I think we shoudl separately discuss making the default TRAF_UPSERT_OPTIMAL  since it's better to make a faster performing solution the default behaviour and still have support for the merge behaviour if needed.